### PR TITLE
fix(button): navigate menu via keyboard with focus

### PR DESCRIPTION
**Pull request by dtsanevmw**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Keyboard navigation in Button Group Menu is not working.

Issue Number: CDE-68

## What is the new behavior?

Keyboard navigation in Button Group Menu is working with arrows

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/angular/src/button/providers/button-group-focus-handler.service.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      case InitialFocus.LAST_ITEM:
+        this.focusLastItem();
+        break;
+      default:
+        this.focusFirstItem();
+        break;
+    }
+  }
+
+  resetButtonsFocus() {
```
Comment (kevinbuhmann) [inline projects/angular/src/button/providers/button-group-focus-handler.service.ts:59]: This method could be marked `private`.

File: projects/angular/src/button/providers/button-group-focus-handler.service.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    this.buttons.forEach(button => {
+      button.blur();
```
Comment (kevinbuhmann) [inline projects/angular/src/button/providers/button-group-focus-handler.service.ts:61]: Why is this needed? I'm just curious because I don't see what it does.

File: projects/angular/src/button/providers/button-group-focus-handler.service.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import { FocusableItem } from '../../utils/focus/focusable-item/focusable-item';
+import { Linkers } from '../../utils/focus/focusable-item/linkers';
+import { InitialFocus } from './button-group-focus.enum';
+
+@Injectable()
+export class ButtonGroupFocusHandler {
+  private initialFocus: InitialFocus = InitialFocus.FIRST_ITEM;
+  private _unlistenFuncs: (() => void)[] = [];
+  private _menuOpened: boolean;
+  private buttons: FocusableItem[];
```
Comment (kevinbuhmann) [inline projects/angular/src/button/providers/button-group-focus-handler.service.ts:20]: `private` members are generally put after `public` members of the same kind. So these should go after the `public` properties.

File: projects/angular/src/button/providers/button-group-focus-handler.service.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+  constructor(
+    private focusService: FocusService,
+    private toggleService: ClrPopoverToggleService,
+    private renderer: Renderer2
+  ) {}
+
+  setInitialFocus(initialFocus: InitialFocus) {
+    this.initialFocus = initialFocus;
+  }
```
Comment (kevinbuhmann) [inline projects/angular/src/button/providers/button-group-focus-handler.service.ts:43]: This could be removed in favor of just making the property `public`. If there is no logic here, we don't need a set method.

File: projects/angular/src/button/button-group/button-group.ts [hunk 5]
```diff
@@ -4,39 +4,58 @@
...
   templateUrl: 'button-group.html',
-  providers: [ButtonInGroupService, ClrDestroyService],
+  providers: [ButtonInGroupService, ClrDestroyService, BUTTON_GROUP_FOCUS_HANDLER_PROVIDER, FOCUS_SERVICE_PROVIDER],
   hostDirectives: [PopoverHostDirective],
   host: { '[class.btn-group]': 'true' },
 })
-export class ClrButtonGroup implements AfterContentInit {
+export class ClrButtonGroup implements AfterContentInit, AfterViewInit {
+  @ViewChild('menuToggle') menuToggle: ElementRef;
+  @ViewChild('menu') menu: ElementRef;
```
Comment (kevinbuhmann) [inline projects/angular/src/button/button-group/button-group.ts:44]: These properties should be marked `private`. And using `ElementRef<HTMLElement>` should be better for type safety.

File: projects/angular/src/button/button-group/button-group.ts [hunk 6]
```diff
@@ -136,4 +166,14 @@ export class ClrButtonGroup implements AfterContentInit {
 
     this.popoverPosition = (ClrPopoverPositions as Record<string, any>)[this._menuPosition];
   }
+
+  private handleFocusOnMenuOpen() {
+    if (this.menuButtons.length) {
+      this.toggleService.openChange.pipe(delay(0), takeUntil(this.destroy$)).subscribe(isOpened => {
+        this.focusHandler.menuToggle = this.menuToggle?.nativeElement;
+        this.focusHandler.menu = this.menu?.nativeElement;
+        this.focusHandler.menuOpened = isOpened;
```
Comment (kevinbuhmann) [inline projects/angular/src/button/button-group/button-group.ts:175]: This `focusHandler.menuOpened` setter has a side effect, and this code does some implicitly-required setup by setting the `menu` and `menuToggle` elements.

I think a better way would be to explicitly call the `focusHandler.initialize` method if the menu is open and pass the elements. This makes the code more explicit about what is happening and what is required.

The `focusHandler.menuOpened` property is not needed at all after this change.

```typescript
if (isOpened) {
  this.focusHandler.initialize({
    menu: this.menu.nativeElement,
    menuToggle: this.menuToggle.nativeElement,
  });
}
```

File: projects/angular/src/button/providers/button-group-focus.enum.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2016-2023 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+export enum InitialFocus {
+  FIRST_ITEM = 'first', // Focus first item
+  LAST_ITEM = 'last', // Focus last item
```
Comment (kevinbuhmann) [inline projects/angular/src/button/providers/button-group-focus.enum.ts:9]: Nit: I don't think these comments are necessary. It's obvious what these enum values mean.

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]
Great work! I have a few comments on improving the code, though.

I worked through my suggested changes myself to make sure everything would work, so I went ahead and pushed my changes to a different branch. You can refer to this commit along with my comments. If you want, you can also cherry-pick my fixup commit. It's up to you!

https://github.com/vmware-clarity/ng-clarity/commits/kevin/cde-68-review

File: projects/angular/src/button/providers/button-group-focus-handler.service.ts (hunk 2 above)
Comment (dtsanevmw) [reply projects/angular/src/button/providers/button-group-focus-handler.service.ts:61]: We're setting "tabindex" to "-1" on blur, so we can have only one with 0 while navigating trough items and that's just resetting all buttons to 0. Using it two times, first when we are initialising the buttons trough link function because they don't have tabindex set to something by default and after closing the menu so they can go back to 0 again.

**Review by kevinbuhmann — APPROVED** [review]
